packages/flutter_tools/lib/src/android/apk_locator.dart

```diff
@@ -0,0 +1,137 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../base/file_system.dart';
+import '../base/utils.dart';
+import '../build_info.dart';
+
+/// A service to locate APK files output by Gradle.
+class ApkLocator {
+  ApkLocator({
+    @required FileSystem fileSystem,
+  }) : _fileSystem = fileSystem;
+
+  final FileSystem _fileSystem;
+
+  /// Retreive a list of candidate bunldes in [root].
+  ///
+  /// If more than one APK is located, the returned list is in order of
+  /// decreasing likeliness to be correct.
+  ///
+  /// Only files that match the provided file extension are considered.
+  List<File> locate(Directory apkDirectory, {
+    Set<String> excludePaths = const <String>{},
+    @required AndroidBuildInfo androidBuildInfo,
+  }) {
+    // First check if the tool can find the exact file.
+    final List<File> exactFiles = <File>[];
+    final BuildInfo buildInfo = androidBuildInfo.buildInfo;
+    final String modeName = camelCase(buildInfo.modeName);
+
+    for (final String fileName in _apkFilesNamesFor(androidBuildInfo)) {
+      final File apkFile = apkDirectory
+        .childDirectory(modeName)
+        .childFile(fileName);
+      if (apkFile.existsSync()) {
+        exactFiles.add(apkFile);
+        continue;
+      }
+
+      if (buildInfo.flavor != null) {
+        // Android Studio Gradle plugin v3 adds flavor to path.
+        final File apkFile = apkDirectory
+          .childDirectory(buildInfo.flavor)
+          .childDirectory(modeName)
+          .childFile(fileName);
+        if (apkFile.existsSync()) {
+          exactFiles.add(apkFile);
+          continue;
+        }
+      }
+    }
```

Comment:
If we could just compare by the APK's last modified timestamp, then we could remove this code. That would be great. 🎉

Reply:
Theres a few scenarios where that  doesn't work, so we need the fallback code:


1. User builds free-flavor and runs. Everything works.
2. User builds paid-flavor and runs. Everything works.
3. User switches back to free-flavor without code changes. Because Gradle has everything correctly cached, the last modified timestamp of the free APK is before the paid APK. Paid is incorrectly selected.

Reply:
I cc'ed you in a Gradle chat - it looks like we might be able to control the filename in Gradle directly.

Reply:
That is great!